### ISSUE TITLE
Existing S3 directory with eggs/wheels not loading

### ISSUE THREAD

**Issue by josepablog** [stevearc/pypicloud#196]
Hello,

We have a s3 "directory" with many eggs and wheels and we would like to use pypicloud to serve them.  I tried using the current directory as a server.ini, but when I rebuild the cache I get an error:

ERROR 2018-11-05 11:20:34,486 [pypicloud.views] can't compare offset-naive and offset-aware datetimes

Any thoughs on how I can use this directory? 

Thanks,
Jose

**Comment by josepablog** [comment]
Ok, this happens when there are two files with the same name in different subdirectories. I guess it is trying to choose one of them and the date format is incorrect? 

Anything I can do from my side to fix this? (Besides deleting files)

**Comment by stevearc** [comment]
Oh boy, I've seen this before. It usually happens when a database loads a timestamp object and assigns it a timezone, then we compare it against a naive python datetime with no timezone. What cache are you using? I've seen this before with the DynamoDB backend but I thought we fixed it. The output of a pip freeze might also help me reproduce it after I see what the DB is.

**Comment by stevearc** [comment]
Oh, also a stacktrace would be great so I can see exactly which comparison is breaking

**Comment by josepablog** [comment]
`ERROR 2018-11-05 21:00:03,652 [pypicloud.views] can't compare offset-naive and offset-aware datetimes
Traceback (most recent call last):
  File "//anaconda/envs/py27/lib/python2.7/site-packages/pyramid/tweens.py", line 41, in excview_tween
    response = handler(request)
  File "//anaconda/envs/py27/lib/python2.7/site-packages/pyramid/router.py", line 148, in handle_request
    registry, request, context, context_iface, view_name
  File "//anaconda/envs/py27/lib/python2.7/site-packages/pyramid/view.py", line 657, in _call_view
    response = view_callable(context, request)
  File "//anaconda/envs/py27/lib/python2.7/site-packages/pyramid/config/views.py", line 188, in attr_view
    return view(context, request)
  File "//anaconda/envs/py27/lib/python2.7/site-packages/pyramid/config/views.py", line 214, in predicate_wrapper
    return view(context, request)
  File "//anaconda/envs/py27/lib/python2.7/site-packages/pyramid/viewderivers.py", line 325, in secured_view
    return view(context, request)
  File "//anaconda/envs/py27/lib/python2.7/site-packages/pyramid/viewderivers.py", line 436, in rendered_view
    result = view(context, request)
  File "//anaconda/envs/py27/lib/python2.7/site-packages/pyramid/viewderivers.py", line 116, in _class_requestonly_view
    response = getattr(inst, attr)()
  File "//anaconda/envs/py27/lib/python2.7/site-packages/pypicloud/views/admin.py", line 25, in rebuild_package_list
    self.request.db.reload_from_storage()
  File "//anaconda/envs/py27/lib/python2.7/site-packages/pypicloud/cache/sql.py", line 272, in reload_from_storage
    return super(SQLCache, self).reload_from_storage(clear)
  File "//anaconda/envs/py27/lib/python2.7/site-packages/pypicloud/cache/base.py", line 76, in reload_from_storage
    self.save(pkg)
  File "//anaconda/envs/py27/lib/python2.7/site-packages/pypicloud/cache/sql.py", line 268, in save
    self.db.merge(package)
  File "//anaconda/envs/py27/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 1910, in merge
    self._autoflush()
  File "//anaconda/envs/py27/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 1434, in _autoflush
    self.flush()
  File "//anaconda/envs/py27/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 2254, in flush
    self._flush(objects)
  File "//anaconda/envs/py27/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 2380, in _flush
    transaction.rollback(_capture_exception=True)
  File "//anaconda/envs/py27/lib/python2.7/site-packages/sqlalchemy/util/langhelpers.py", line 66, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "//anaconda/envs/py27/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 2344, in _flush
    flush_context.execute()
  File "//anaconda/envs/py27/lib/python2.7/site-packages/sqlalchemy/orm/unitofwork.py", line 391, in execute
    rec.execute(self)
  File "//anaconda/envs/py27/lib/python2.7/site-packages/sqlalchemy/orm/unitofwork.py", line 556, in execute
    uow
  File "//anaconda/envs/py27/lib/python2.7/site-packages/sqlalchemy/orm/persistence.py", line 177, in save_obj
    mapper, table, update)
  File "//anaconda/envs/py27/lib/python2.7/site-packages/sqlalchemy/orm/persistence.py", line 684, in _emit_update_statements
    lambda rec: (
  File "//anaconda/envs/py27/lib/python2.7/site-packages/sqlalchemy/orm/persistence.py", line 492, in _collect_update_commands
    value, state.committed_state[propkey]) is not True:
  File "//anaconda/envs/py27/lib/python2.7/site-packages/sqlalchemy/sql/type_api.py", line 364, in compare_values
    return x == y
TypeError: can't compare offset-naive and offset-aware datetimes`

**Comment by josepablog** [comment]
Thank you for your help! I pasted the stack trace above. 
This is what my config file has:
`db.url = sqlite:///%(here)s/db.sqlite`

Should I change it to dynamo to see if helps?

**Comment by stevearc** [comment]
Nah, it should be possible to make it work with SQL. I see kinda what's happening, though I don't know exactly the root cause yet. While I'm working on that, you could try setting [db.graceful_reload = true](https://pypicloud.readthedocs.io/en/latest/topics/cache.html#db-graceful-reload). That will trigger a slightly different code path that *may* not have the same issue

**Comment by josepablog** [comment]
Ohhh thank you so much Steve! That worked.

**Comment by josepablog** [comment]
This is a great project, BTW.

**Comment by stevearc** [comment]
Glad it worked! From looking at the error I would guess that one of two things is happening.

1) The boto API is returning mixed datetime objects from S3. You could check this by logging the tzinfo as we pull it from S3 [on line 129](https://github.com/stevearc/pypicloud/blob/42e10c1fab8e72633d3f2b9c4e1b22675cb531b6/pypicloud/storage/s3.py#L129)
```
LOG.debug("Last Modified: %s %s", name, obj.last_modified.tzinfo)
```
2) The SQLAlchemy merge operation converts some of the datetimes from one format to the other and the comparison breaks mid-batch. This would be a lot more difficult to track down. We'd have to verify that the DB is empty, that all the datetimes being passed in are the same format, and narrow it down to a minimal set of Package objects that causes a repro. Then we could pass it upstream to SQLAlchemy.

If you're done and don't want to bother to debug the original issue, that's totally fine. I'm just going to leave this information here so if anyone else has the same problem they'll see where we left off.

**Comment by stevearc** [comment]
datetime comparison crashes should all be fixed now #266
